### PR TITLE
Issue #3 : Enhancements With Distraction Tracking, Data Storing And Visualization And CSV Export 

### PR DESCRIPTION
**Pull request by SpreadSheets600**
### Summary:
This PR includes significant improvements to the Pomodoro timer functionality, distraction protection, and user experience. It focuses on refining fullscreen handling, improving the protection mechanisms, and ensuring smooth operation during Pomodoro sessions. Additionally, logging and error handling were improved, and the code was refactored for better performance.

**Key Changes:**

1. **Improved Fullscreen Management:**
   - Refined the fullscreen handling for the browser window. Instead of the previously failed `ensure_fullscreen` method, `pyautogui.press('f11')` is now directly used to enter and exit fullscreen mode, ensuring that the user's focus is maintained during the session.
   - Added handling for both minimizing and maximizing the Flask window when starting and completing the Pomodoro session using `pygetwindow`. This enhances the overall user experience, ensuring that the Flask UI is minimized during focus and restored when the session completes.

2. **Mouse Boundary Enforcement:**
   - Implemented a mechanism to enforce mouse boundaries during Pomodoro sessions using the `enforce_mouse_boundaries` method in the `FocusProtection` class. The mouse position is restricted to prevent distractions caused by accidental mouse movement to the screen's top or bottom.

3. **Blocking Distractions via Keyboard Keys:**
   - Expanded the distraction blocking mechanism by defining a set of blocked keys (`CONFIG["BLOCKED_KEYS"]`). This ensures that critical keys like `alt`, `tab`, `ctrl`, and `windows` are disabled during the focus sessions, preventing unwanted distractions.

4. **Threading for Pomodoro Flow:**
   - Refactored the `pomodoro_flow` function to run in a separate thread, ensuring that the main application remains responsive while the Pomodoro session is active.
   - Improved threading management for handling protection features (`block_distractions`), with proper activation and cleanup of mouse threads to avoid resource leaks.

5. **Exit Combo for Safe Exit:**
   - Introduced an exit combo (`ctrl+shift+q`) that allows the user to safely exit the focus protection if needed before the session is complete. This gives users more control without compromising the protection mechanisms.

6. **Error Handling and Logging:**
   - Improved error handling around the protection mechanisms to ensure the mouse enforcement and key blocking are properly initialized and cleaned up, even if exceptions occur.
   - Logged key events to monitor any issues with blocked keys or fullscreen toggling.

7. **Miscellaneous:**
   - Added configurable delay for mouse boundary enforcement (`CONFIG["MOUSE_ENFORCE_DELAY"]`) to fine-tune the responsiveness of the mouse restrictions.
   - Improved the structure of the app by cleaning up unused or failed methods (e.g., commented out the previously failed `ensure_fullscreen` function for future reference).

8. **Data Visualization:**
   - Added the ability to visualize focus session data using bar graphs and line charts for daily, weekly, and monthly breakdowns. This feature helps users track their focus performance over time.
   - Graphs display total focus time, number of distractions, and time spent in fullscreen mode, providing a clear picture of session performance.
   - Integrated CSV data extraction, allowing users to export session data for external analysis or record-keeping.
   
9. **Database Integration:**
   - Introduced a database storage system to track session data, including total focus duration, distractions, and fullscreen events. This allows users to store and retrieve session data for historical analysis.
   - Implemented a mechanism to save data automatically after each session, supporting future features like data recovery and session analytics.

**Testing:**
- Manually tested the Pomodoro session flow with different durations for focus and break periods.
- Verified that mouse movement is constrained, fullscreen mode is properly managed, and distraction-blocking keys are effectively disabled during the session.
- Checked the proper behavior of the exit combo to safely quit focus protection.

------
### TODO
1. Improve distraction tracking
2. Add more data visualization method
3. Improve the web UI specifically for the data
4. Lastly, add a timer while the focus time is on ( improvement, but not necessary )
5. more user customizations

### CODE REVIEW

File: amine/webapp.py [hunk 1]
```diff
@@ -1,183 +1,521 @@
...
+import pyautogui
 import threading
 import webbrowser
-from datetime import datetime, timedelta
-import pyautogui
-import keyboard
-import time
-import sys
+
+import matplotlib
```
Comment (datavorous) [inline amine/webapp.py:11]: Please use poetry to add this libraires in your venv, they are not updated in the .toml file

File: amine_data.db
Comment (datavorous) [inline amine_data.db:1]: please delete this file

File: amine/webapp.py [hunk 2]
```diff
@@ -1,183 +1,521 @@
...
+    plt.grid(True)
+
+    img = BytesIO()
+    plt.savefig(img, format="png")
+    img.seek(0)
+    plt.close()
+    return send_file(img, mimetype="image/png")
+
+
+@app.route("/distraction_heatmap")
```
Comment (datavorous) [inline amine/webapp.py:356]: Heatmap is not being rendered.

**Review by datavorous — CHANGES_REQUESTED** [review]
1. use poetry 
2. remove the .db file
3. heatmaps are not being rendered

**Comment by datavorous** [comment]
Please rebase your code with the latest version. A separate PR has been merged.

**Comment by datavorous** [comment]
> Please rebase your code with the latest version. A separate PR has been merged.

Please execute this and verify.

**Comment by SpreadSheets600** [comment]
> > Please rebase your code with the latest version. A separate PR has been merged.
> 
> Please execute this and verify. @SpreadSheets600 

Okay I will do it :)

**Comment by SpreadSheets600** [comment]
@datavorous Done Added and Updated It with the Graphs
PS : I Was thinking of separate graphs for each, exe mode and website mode what do you say ?

**Pull request merged**